includes/class-stats.php

```diff
@@ -59,22 +60,115 @@ public function migrate() {
 				"CREATE TABLE {$wpdb->wpjm_stats} (
 				`id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
 				`date` date NOT NULL,
-				`post_id` bigint(20) DEFAULT NULL,
+				`post_id` bigint(20) DEFAULT 0,
 				`name` varchar(255) NOT NULL,
 				`group` varchar(255) DEFAULT '',
 				`count` bigint(20) unsigned not null default 1,
 				PRIMARY KEY (`id`),
-				INDEX `idx_wpjm_stats_name_date_group`  (`name`, `date`, `group`)
+				INDEX `idx_wpjm_stats_name_date_group`  (`name`, `date`, `group`),
+				UNIQUE INDEX `idx_wpjm_stats_name_date_group_post_id`  (`name`, `date`, `group`, `post_id`)
 			) {$collate}",
 			]
 		);
 	}
 
+	/**
+	 * Log a stat into the db.
+	 *
+	 * @param string $name         The stat name.
+	 * @param string $group        The group this stat belongs to.
+	 * @param int    $post_id      The post_id this stat belongs to.
+	 * @param int    $increment_by The amount to increment the stat by.
+	 *
+	 * @return bool
+	 */
+	public function log_stat( string $name, string $group = '', int $post_id = 0, int $increment_by = 1 ) {
+		global $wpdb;
+
+		if (
+			strlen( $name ) > 255 ||
+			strlen( $group ) > 255 ||
+			! is_numeric( $post_id ) ||
+			! is_numeric( $increment_by ) ) {
+			return false;
+		}
+
+		$date_today = gmdate( 'Y-m-d' );
+
+		// phpcs:disable WordPress.DB.DirectDatabaseQuery.DirectQuery
+		$result = $wpdb->query(
+			$wpdb->prepare(
+				"INSERT INTO {$wpdb->wpjm_stats} " .
+				'(`name`, `date`, `group`, `post_id`, `count` ) ' .
+				'VALUES (%s, %s, %s, %d, %d) ' .
+				'ON DUPLICATE KEY UPDATE `count` = `count` + %d',
+				$name,
+				$date_today,
+				$group,
+				$post_id,
+				$increment_by,
+				$increment_by
+			)
+		);
+
+		if ( false === $result ) {
+			return false;
+		}
+
+		$cache_key = $this->get_cache_key_for_stat( $name, $group, $post_id );
+
+		wp_cache_delete( $cache_key, 'wpjm_stats' );
+
+		return true;
+	}
+
+	/**
+	 * Get a cache key for a stat.
+	 *
+	 * @param string $name    The name.
+	 * @param string $group   The optional group.
+	 * @param int    $post_id The optional post id.
+	 *
+	 * @return string
+	 */
+	private function get_cache_key_for_stat( string $name, string $group = '', int $post_id = 0 ) {
+		$hash = md5( "{$name}_{$group}_{$post_id}" );
+		return "wpjm_stat_{$hash}";
+	}
+
 	/**
 	 * Perform plugin activation-related stats actions.
 	 *
 	 * @return void
 	 */
 	public function activate() {
 	}
+
+	/**
+	 * Run any hooks related to stats.
+	 *
+	 * @return void
+	 */
+	private function hook() {
+		add_action( 'wp', [ $this, 'maybe_log_listing_view' ] );
+	}
+
+	/**
+	 * Log a (non-unique) listing page view.
+	 *
+	 * @return void
+	 */
+	public function maybe_log_listing_view() {
+		if ( is_admin() || ( defined( 'DOING_AJAX' ) && DOING_AJAX ) ) {
+			return;
+		}
+
+		$post_id   = absint( get_queried_object_id() );
+		$post_type = get_post_type( $post_id );
+		if ( 'job_listing' !== $post_type ) {
```

Comment:
Reminder to use the constants, in case you decide to NOT use the `job_manager_single_job_content` filter:

```suggestion
		if ( \WP_Job_Manager_Post_Types::PT_LISTING !== $post_type ) {
```